### ISSUE TITLE
Concept Samjhte Hain
Ye tool machine learning ka sabse basic form hai — Linear Regression — jo NumPy se scratch mein banaya ja sakta hai (bina sklearn/TensorFlow ke).
Idea simple hai: Agar hamare paas pichle kuch saalon ka data ho —
Kitni baarish hui (mm mein)
Mitti mein kitna paani/moisture tha
Temperature kya thi
Fertilizer kitna use hua
Us saal fasal (crop) kitni hui (yield, jaise maund/acre)
...to hum ek formula seekh sakte hain jo in factors se yield predict kare. Agle saal agar mausam ka andaza ho, to farmer pehle se jaan sakta hai uski fasal kitni hogi — planning, loan, aur resources better manage kar sakta hai.


### ISSUE THREAD

**Issue by muattarnaz7-svg** [muattarnaz7-svg/Crop-Yield-Predictor-for-Local-Farmers-apne-region-ki-soil-rain-data-se-chhoti-scale-prediction-tool#1]
# -*- coding: utf-8 -*-
"""
Crop Yield Predictor for Local Farmers
------------------------------------------
A small-scale crop yield prediction tool built using pure NumPy
(no external ML libraries like scikit-learn/TensorFlow needed).

Uses Multiple Linear Regression, solved via the Normal Equation
(matrix algebra), to predict crop yield from:
    - Rainfall (mm)
    - Soil moisture (%)
    - Average temperature (deg C)
    - Fertilizer used (kg/acre)

A farmer (or local agri-officer) can collect a few years of their
own field data, plug it in here, and get a working predictor
specific to THEIR land -- not a generic national model.

Author: Built with Claude
"""

import numpy as np
import csv

# ----------------------------------------------------------------------
# 1. SAMPLE DATA  (replace with real records from your own field/region)
# ----------------------------------------------------------------------
def generate_sample_data(seed=7):
    """
    Simulates 12 seasons of farm data for one field.
    Columns: rainfall(mm), soil_moisture(%), temperature(C), fertilizer(kg/acre)
    Target: yield (maund/acre)

    Replace this with real historical records for actual use.
    """
    rng = np.random.default_rng(seed)

    n_seasons = 12
    rainfall = rng.uniform(200, 800, n_seasons)          # mm per season
    soil_moisture = rng.uniform(15, 45, n_seasons)        # percent
    temperature = rng.uniform(20, 38, n_seasons)          # deg C
    fertilizer = rng.uniform(20, 100, n_seasons)          # kg/acre

    # Simulate a realistic (but made-up) relationship + some randomness,
    # like real farm outcomes are never perfectly predictable.
    yield_maund = (
        0.04 * rainfall
        + 0.6 * soil_moisture
        + 0.3 * fertilizer
        - 0.5 * np.abs(temperature - 28)   # crops like moderate temp best
        + rng.normal(0, 2, n_seasons)      # random noise
        + 10
    )
    yield_maund = np.clip(yield_maund, 5, None)  # yield can't be negative

    X = np.column_stack([rainfall, soil_moisture, temperature, fertilizer])
    y = yield_maund

    feature_names = ["Rainfall(mm)", "Soil Moisture(%)", "Temperature(C)", "Fertilizer(kg/acre)"]
    return X, y, feature_names

# ----------------------------------------------------------------------
# 2. LOAD REAL DATA FROM CSV  (use this once you have real farm records)
# ----------------------------------------------------------------------
def load_csv_data(path):
    """
    Expects a CSV with header row:
    rainfall,soil_moisture,temperature,fertilizer,yield

    Returns X (features), y (target), feature_names
    """
    rows = []
    with open(path, "r", encoding="utf-8") as f:
        reader = csv.reader(f)
        header = next(reader)
        for row in reader:
            rows.append([float(val) for val in row])

    data = np.array(rows)
    X = data[:, :-1]
    y = data[:, -1]
    feature_names = header[:-1]
    return X, y, feature_names

# ----------------------------------------------------------------------
# 3. NORMALIZE FEATURES  (important: rainfall is in 100s, moisture in %)
# ----------------------------------------------------------------------
def normalize_features(X):
    """
    Scales each feature to have mean 0, std 1.
    This prevents large-scale features (like rainfall in mm) from
    dominating small-scale ones (like temperature) during training.

    Returns normalized X, plus mean/std so new predictions can be
    normalized the same way.
    """
    mean = X.mean(axis=0)
    std = X.std(axis=0)
    std[std == 0] = 1  # avoid divide-by-zero for constant columns
    X_norm = (X - mean) / std
    return X_norm, mean, std

# ----------------------------------------------------------------------
# 4. TRAIN LINEAR REGRESSION VIA THE NORMAL EQUATION
# ----------------------------------------------------------------------
def train_linear_regression(X_norm, y):
    """
    Solves for regression weights using the Normal Equation:
        theta = (X^T X)^-1 X^T y

    This is exact (no iterative training needed) and works great
    for small datasets like a single farmer's seasonal records.
    """
    n_samples = X_norm.shape[0]

    # Add a column of 1s for the bias/intercept term
    X_b = np.column_stack([np.ones(n_samples), X_norm])

    # Normal equation (using pinv = pseudo-inverse for numerical safety)
    theta = np.linalg.pinv(X_b.T @ X_b) @ X_b.T @ y

    return theta  # theta[0] = intercept, theta[1:] = feature weights

# ----------------------------------------------------------------------
# 5. PREDICT YIELD FOR NEW CONDITIONS
# ----------------------------------------------------------------------
def predict_yield(theta, X_new, mean, std):
    """
    Predicts yield for new input conditions.
    X_new: 1D array like [rainfall, soil_moisture, temperature, fertilizer]
    """
    X_new = np.array(X_new, dtype=float)
    X_new_norm = (X_new - mean) / std
    X_b = np.concatenate([[1], X_new_norm])
    prediction = X_b @ theta
    return float(prediction)

# ----------------------------------------------------------------------
# 6. EVALUATE MODEL ACCURACY  (R-squared)
# ----------------------------------------------------------------------
def evaluate_model(theta, X_norm, y):
    """
    Calculates R^2 (coefficient of determination):
    1.0 = perfect fit, 0.0 = no better than guessing the average.

    Also returns predictions and residuals (errors) for inspection.
    """
    n_samples = X_norm.shape[0]
    X_b = np.column_stack([np.ones(n_samples), X_norm])
    predictions = X_b @ theta

    residuals = y - predictions
    ss_residual = np.sum(residuals ** 2)
    ss_total = np.sum((y - y.mean()) ** 2)

    r_squared = 1 - (ss_residual / ss_total) if ss_total != 0 else 0
    mae = np.mean(np.abs(residuals))          # average error in maund/acre
    rmse = np.sqrt(np.mean(residuals ** 2))   # penalizes big misses more

    return {
        "r_squared": round(float(r_squared), 3),
        "mae": round(float(mae), 2),
        "rmse": round(float(rmse), 2),
        "predictions": predictions,
        "residuals": residuals,
    }

# ----------------------------------------------------------------------
# 7. FEATURE IMPORTANCE  (which factor affects yield the most?)
# ----------------------------------------------------------------------
def feature_importance(theta, feature_names):
    """
    Since features were normalized before training, the size of each
    weight directly reflects how much influence that factor has.
    Larger absolute weight = bigger impact on yield.
    """
    weights = theta[1:]  # skip intercept
    order = np.argsort(-np.abs(weights))

    results = []
    for idx in order:
        results.append((feature_names[idx], round(float(weights[idx]), 3)))
    return results

# ----------------------------------------------------------------------
# 8. REPORT PRINTER
# ----------------------------------------------------------------------
def print_report(X, y, feature_names, theta, mean, std, eval_results):
    print("=" * 55)
    print("        CROP YIELD PREDICTOR - MODEL REPORT")
    print("=" * 55)
    print(f"Training seasons used : {len(y)}")
    print(f"Model accuracy (R^2)  : {eval_results['r_squared']}  (closer to 1.0 = better)")
    print(f"Avg error (MAE)       : {eval_results['mae']} maund/acre")
    print(f"RMSE                  : {eval_results['rmse']} maund/acre")
    print("-" * 55)

    print("FACTOR IMPORTANCE (biggest impact on yield first):")
    for name, weight in feature_importance(theta, feature_names):
        direction = "increases" if weight > 0 else "decreases"
        print(f"  {name:<22}: weight = {weight:>7}  ({direction} yield)")

    print("-" * 55)
    print("PAST SEASONS: actual vs predicted yield")
    print(f"{'Season':<10}{'Actual':<12}{'Predicted':<12}{'Error':<10}")
    for i in range(len(y)):
        print(f"{i+1:<10}{round(y[i],2):<12}{round(eval_results['predictions'][i],2):<12}{round(eval_results['residuals'][i],2):<10}")
    print("=" * 55)

# ----------------------------------------------------------------------
# 9. MAIN
# ----------------------------------------------------------------------
if __name__ == "__main__":
    # ---- Option A: use simulated sample data (works out of the box) ----
    X, y, feature_names = generate_sample_data()

    # ---- Option B: load your own region's real farm data instead ----
    # X, y, feature_names = load_csv_data("my_farm_data.csv")

    X_norm, mean, std = normalize_features(X)
    theta = train_linear_regression(X_norm, y)
    eval_results = evaluate_model(theta, X_norm, y)

    print_report(X, y, feature_names, theta, mean, std, eval_results)

    # ---- Example: predict yield for a NEW upcoming season ----
    print("\nPREDICTION FOR A NEW SEASON:")
    new_conditions = [450, 30, 29, 60]   # rainfall, soil_moisture, temp, fertilizer
    predicted = predict_yield(theta, new_conditions, mean, std)
    print(f"Input  -> Rainfall: {new_conditions[0]}mm, Soil Moisture: {new_conditions[1]}%, "
          f"Temp: {new_conditions[2]}C, Fertilizer: {new_conditions[3]}kg/acre")
    print(f"Predicted Yield -> {round(predicted, 2)} maund/acre")
